.design/requirements-575.md

```diff
@@ -0,0 +1,269 @@
+
+```java
+@Component(service = WebSocketEndpointExample.class, scope = ServiceScope.PROTOTYPE)
+@WebSocketWhiteboardServletPattern("/ws/work/{targetId}") // See "WebSocket pattern description"
+public class WebSocketEndpointExample {
+    @OnOpen
+    public void onOpen(
+        @PathParam("targetId") String targetId,
+        Session session) {
+        // WebSocket stream is opened
+    }
+
+    @OnClose
+    public void onClose(
+        @PathParam("targetId") String targetId,
+        Session session,
+        CloseReason reason) {
+        // WebSocket stream is closed
+    }
+
+    @OnError
+    public void onError(
+        @PathParam("targetId") String targetId,
+        Session session,
+        Throwable error) {
+        // Error handling the WebSocket stream
+    }
+
+    @OnMessage
+    public void onMessage(
+        @PathParam("targetId") String targetId,
+        String message,
+        Session session) {
+        // Process incoming message
+		// The "message" argument can also be of type Reader, Decoder.Text
+		// or Decoder.TextStream.
+		// See @OnMessage documentation for more details.
+    }
+
+	@OnMessage
+    public void onMessage(
+        @PathParam("targetId") String targetId,
+        byte[] message,
+        Session session) {
+        // Process binary message
+		// The "message" argument can also be of type ByteBuffer, InputStream,
+		// Decoder.Binary or Decoder.BinaryStream.
+		// See @OnMessage documentation for more details.
+    }
+
+    @OnMessage
+    public void onPong(
+        @PathParam("targetId") String targetId,
+        PongMessage message,
+        Session session) {
+        // Process pong message
+    }
+}
+```
+
+#### WebSocket pattern description
+
+In a Jakarta environment, we would annotate this class with `@ServerEndpoint("/ws/work/{targetId}")` and register it in a Jakarta WebSocket-ready HTTP server.
+
+In OSGi, we might want to use annotations similar to the ones in the Jarkata Servlet whiteboard specification to provide more control, *e.g.* `@WebSocketWhiteboardServletPattern("/ws/work/{targetId}")`.
+
+#### WebSocket subprotocols
+
+The [WebSocket specification (RFC 6455)](https://datatracker.ietf.org/doc/html/rfc6455#section-1.9) indicates that clients can request, during the handshake, can give a list of subprotocols names they expect, in order of preference.
+If a list of subprotocols is specified, the server must use and return the subprotocol it selected from that list.
+
+RFC 6455 doesn't explicitly specifies what happens if the server doesn't support any of the client subprotocols.
+
+The list of subprotocols supported by the OSGi WebSocket Whiteboard server endpoint can be given using a service property of type `string[]`.
+A class annotation, *e.g.* `WebSocketWhiteboardSubprotocols`, could be used to ease the declaration of such list.
+
+#### Service scope
+
+The Jakarta WebSocket specification expects one instance of WebSocket handler per session.
```

Comment:
This is not true. The Specification states that a server is usually one instance, a client is an instance per connection if not otherwise configured.

Servers *must* be able to handle different client request with different sessions!